scheduling-utils/src/handshake/server.rs

```diff
@@ -0,0 +1,279 @@
+        let sent =
+            socket::sendmsg::<UnixAddr>(stream.as_raw_fd(), &iov, &cmsgs, MsgFlags::empty(), None)
+                .map_err(std::io::Error::from)?;
+        debug_assert_eq!(sent, 1);
+
+        Ok(Session {
+            tpu_to_pack,
+            progress_tracker,
+            pack_to_worker,
+            worker_to_pack,
+        })
+    }
+
+    fn create_producer<T>(size: usize) -> Result<(File, shaq::Producer<T>), ShaqError> {
+        let file = Self::create_shmem(size)?;
+        let queue = shaq::Producer::create(&file, size)?;
+
+        Ok((file, queue))
+    }
+
+    fn create_consumer(
+        size: usize,
+    ) -> Result<(File, shaq::Consumer<PackToWorkerMessage>), ShaqError> {
+        let file = Self::create_shmem(size)?;
+        let queue = shaq::Consumer::create(&file, size)?;
+
+        Ok((file, queue))
+    }
+
+    fn create_shmem(size: usize) -> Result<File, ShaqError> {
+        let size_t = i64::try_from(size).map_err(|_| ShaqError::InvalidBufferSize)?;
+
+        // Create a memory backed file.
+        #[cfg(any(
+            target_os = "linux",
+            target_os = "l4re",
+            target_os = "android",
+            target_os = "emscripten"
+        ))]
+        let file = Self::memfd_create()?;
+        #[cfg(not(any(
+            target_os = "linux",
+            target_os = "l4re",
+            target_os = "android",
+            target_os = "emscripten"
+        )))]
+        let file = Self::shmem_create()?;
+
+        // Resize to the requested size.
+        let ret = unsafe { libc::ftruncate64(file.as_raw_fd(), size_t) };
+        if ret == -1 {
+            return Err(ShaqError::Io(std::io::Error::last_os_error()));
+        }
+        debug_assert_eq!(ret, 0);
+
+        Ok(file)
+    }
+
+    #[cfg(any(
+        target_os = "linux",
+        target_os = "l4re",
+        target_os = "android",
+        target_os = "emscripten"
+    ))]
+    fn memfd_create() -> Result<File, std::io::Error> {
+        unsafe {
+            let ret = libc::memfd_create(c"agave".as_ptr(), 0);
+            if ret == -1 {
+                return Err(std::io::Error::last_os_error());
+            }
+
+            Ok(File::from_raw_fd(ret))
+        }
+    }
+
+    #[cfg(not(any(
+        target_os = "linux",
+        target_os = "l4re",
+        target_os = "android",
+        target_os = "emscripten"
+    )))]
+    fn shmem_create() -> Result<File, std::io::Error> {
+        unsafe {
+            // Clean up the previous link if one exists.
+            let ret = libc::shm_unlink(c"agave".as_ptr());
+            if ret == -1 {
+                let err = std::io::Error::last_os_error();
+                if err.kind() != std::io::ErrorKind::NotFound {
+                    return Err(err);
+                }
+            }
+
+            // Create a new shared memory object.
+            let ret = libc::shm_open(
+                c"agave".as_ptr(),
```

Comment:
From docs I think we need to have it start with a "/" :

> For portable use, a shared memory object should be identified by a
> name of the form /somename; that is, a null-terminated string of
> up to NAME_MAX (i.e., 255) characters consisting of an initial
> slash, followed by one or more characters, none of which are
> slashes.

Reply:
I also think we should probably use a more descriptive name.
If we do something similar in the future to break out other components of agave, we don't want to run into issues where they try to open the same shm region.